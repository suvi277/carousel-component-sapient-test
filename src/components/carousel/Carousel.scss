.carousel-wrapper {
  position: relative;
  margin: 0 auto;
  text-align: center;

  .carousel-items {
    padding: 0;
    margin: 0 -10px;
    display: flex;
    overflow: hidden;
  }

  .carousel-button {
    position: absolute;
    transform: translateY(-50%);
    top: 50%;
    z-index: 1;
    background: url(arrow.svg);
    background-size: cover;
    width: 40px;
    height: 70px;
    border: none;
    color: transparent;

    &.prev {
      left: 0;
      transform: translateY(-50%) scaleX(-1);
    }

    &.next {
      right: 0
    }
  }

  @media screen and (min-width: 1024px) {
    .carousel-button {
      position: static;
      width: auto;
      height: auto;
      transform: none;
      padding: 12px 15px;
      font-size: 14px;
      background: #FF5733 none;
      text-transform: capitalize;
      color: #ffffff;
      border: none;
      cursor: pointer;
  
      &:hover {
        background-color: #FF7133;
      }
  
      &.prev {
        margin-right: 15px;
        transform: none;
      }
    }
  }
}