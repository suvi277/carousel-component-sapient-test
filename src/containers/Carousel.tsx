import React from 'react';
import { connect } from 'react-redux';
import { State } from '../reducers';
import Loader from "react-loader-spinner";
import { getPhotos } from '../actions/photos/photos.action';
import { Photo } from '../types/Photo';
import { useEffect } from 'react';

interface Props {
	status: string;
	photos?: Array<Photo>;
	getPhotos: any;
}

const CarouselContainer: React.FC<Props> = ({ photos = [], status, getPhotos }) => {
	useEffect(() => {
		getPhotos();
	}, [getPhotos]);

	return (
		<div className="container">
			{status === 'loading' ? (
				<Loader type="Grid" color="#00BFFF" height={100} width={100} timeout={3000}/>
			) : (
				status === 'loaded' && (
						<div>Carousel Component</div>
				)
			)}
			{status === 'error' && <div className="error-content">Oops there is an error! Try Again later</div>}
		</div>
	);
};

const mapStateToProps = (state: State) => ({ ...state.photos });

const mapDispatchToProps = (dispatch: any) => ({ getPhotos: () => dispatch(getPhotos()) });

export default connect(mapStateToProps, mapDispatchToProps)(CarouselContainer);
