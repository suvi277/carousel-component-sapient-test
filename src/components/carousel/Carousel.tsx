import React, { useEffect, useState } from 'react';
import PhotoCard from '../photoCard/PhotoCard';
import './Carousel.scss';
import { Photo } from '../../types/Photo';

interface Props {
	photos: Array<Photo>;
}

const CarouselComponent: React.FC<Props> = ({ photos = [] }) => {
	const [ current, setCurrent ] = useState(0);
	const [visibleItems, setVisibleItems] = useState(1);
	
	useEffect(() => {
		setVisibleItemOnScreen();
		window.addEventListener('resize', setVisibleItemOnScreen);
		return () => {
			window.removeEventListener('resize', setVisibleItemOnScreen);
		}
	}, []);

	const setVisibleItemOnScreen = () => {
		let item = 1;
		if (window.innerWidth >= 1024) {
			item = 5;
		} else if (window.innerWidth >= 768) {
			item = 3
		}
		
		setVisibleItems(item);
	}
  	const onPrevClick = () => {
		const resetToVeryBack = current === 0;
		const index = resetToVeryBack ? photos.length - 1 : current - 1;
		setCurrent(index);
	};

	const onNextClick = () => {
		const resetIndex = current === photos.length - 1;

		const index = resetIndex ? 0 : current + 1;
		setCurrent(index);
	}

	const slicePhotos = photos.slice(current, current + visibleItems);
	
  	const carouselItems = slicePhotos.length < visibleItems
    		? [...slicePhotos, ...photos.slice(0, visibleItems - slicePhotos.length) ]
			: slicePhotos;
	
	return (
		<div className="carousel-wrapper">
			<ul className="carousel-items" data-testid="carousel-items">
				{carouselItems.map((photo, index) => <PhotoCard key={index} {...photo}/>)}
			</ul>
			<button data-testid="prev-btn" type="button" onClick={onPrevClick} aria-label="previous" className="carousel-button prev">prev</button>
			<button data-testid="next-btn" type="button" onClick={onNextClick} aria-label="next" className="carousel-button next">next</button>
		</div>
	);
};

export default CarouselComponent;
