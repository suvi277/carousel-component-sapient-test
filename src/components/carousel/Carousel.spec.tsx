import React from 'react';
import { shallow, mount } from 'enzyme';
import Carousel from './Carousel';

describe('<Carousel /> with default props', () => {
	const defaultProps = {
		photos: []
	};
	const wrapper = shallow(<Carousel {...defaultProps} />);
	it('renders as expected', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
