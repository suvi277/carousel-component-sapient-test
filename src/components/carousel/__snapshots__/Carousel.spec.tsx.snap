// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> with default props renders as expected 1`] = `ShallowWrapper {}`;
